addons/mail/models/mail_channel.py

```diff
@@ -238,12 +376,53 @@ def message_get_recipient_values(self, notif_message=None, recipient_ids=None):
             }
         return super(Channel, self).message_get_recipient_values(notif_message=notif_message, recipient_ids=recipient_ids)
 
+    def _extract_values(self, message_type, **kwargs):
+        """ This method is used to compute moderation status before the creation of a message. 
+            For this operation the message's author email address is required.
+            This address is returned with status for other computations.
+        """
+        moderation_status = 'accepted'
+        email = ''
+        if self.moderation and message_type in ['email', 'comment']:
+            author_id = kwargs.get('author_id')
+            if author_id and isinstance(author_id, pycompat.integer_types):
+                author = self.env['res.partner'].browse([author_id])
+                email = author.email
+            elif author_id:
+                email = author_id.email
+            elif self.env.user:
+                email = self.env.user.email
+            else:
+                email_from = kwargs.get('email_from')
+                if email_from:
+                    email = tools.email_split(email_from)[0]
+            if self.env['mail.moderation.email'].search([('email', '=', email), ('status', '=', 'allow')]) or email in self.mapped('moderator_ids.email'):
+                moderation_status = 'accepted'
+            elif self.env['mail.moderation.email'].search([('email', '=', email), ('status', '=', 'ban')]):
+                moderation_status = 'rejected'
+            else:
+                moderation_status = 'pending_moderation'
+        return moderation_status, email
+
     @api.multi
-    @api.returns('self', lambda value: value.id)
-    def message_post(self, **kwargs):
-        # auto pin 'direct_message' channel partner
+    @api.returns('mail.message', lambda value: value.id)
+    def message_post(self, body='', subject=None, message_type='notification', subtype=None, parent_id=False, attachments=None, 
+        content_subtype='html', **kwargs):
+        moderation_status, email= self._extract_values(message_type, **kwargs)
+        if moderation_status == 'rejected':
+            return self.env['mail.message']
+        if self.moderation_notify and message_type == 'email' and moderation_status == 'pending_moderation':
+        # Notifies the message author when his message is pending moderation if required on channel.
+        # The fields "email_from" and "reply_to" are filled in automatically (by the method create in model mail.message).
+            if not subject:
+                subject = ''
+            self.env['mail.message'].create_notification_email({
```

Comment:
Just directly create a mail.mail with state = outgoing. create_notification_email is an unnecessary wrapper.

Reply:
done. i have erase the method create_notification_email and modified accordingly the rest of the code that used it.